### DISCUSSION TITLE
Duplicate typenames

### DISCUSSION THREAD

**Discussion by pavelwitassek** [phpro/soap-client#499] [Q&A] [answered]
Hi,

how can I solve duplicate type names in one request? Example of WSDL is https://ws.gordic.cz/390/Ssl1/Ssl.svc?WSDL and i try to implement Ws-info method, which has multiple duplicate typenames.

Second problem is, that method name is Ws-info and i got warning from line vendor/php-soap/ext-soap-engine/src/Metadata/MethodsParser.php:73, Is possible to extend regex to parsing name and return types?

Example of working SOAP request to Ws-info.
`<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" xmlns:wsa="http://www.w3.org/2005/08/addressing">
    <SOAP-ENV:Header>
        <wsa:Action>http://www.gordic.cz/svc/xrg-ssl/v_1.0.0.0/Ws-info</wsa:Action>
        <wsa:To>https://ws.gordic.cz/390/Ssl1/Ssl.svc</wsa:To>
        <wsa:MessageID>uuid:a5ce3504-4e0f-c564-88d2-22f02999c42f</wsa:MessageID>
        <wsa:ReplyTo>
            <wsa:Address>http://www.w3.org/2005/08/addressing/anonymous</wsa:Address>
        </wsa:ReplyTo>
        <wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" SOAP-ENV:mustUnderstand="1">
            <wsse:UsernameToken>
                <wsse:Username>test</wsse:Username>
                <wsse:Password Type="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText">testovic</wsse:Password>
                <wsse:Nonce EncodingType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#Base64Binary">faSYJllewKq1MT8I0l/GhEYpoopEuBfxObbegU3Y+ws=</wsse:Nonce>
                <wsu:Created xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">2024-01-21T21:18:49Z</wsu:Created>
            </wsse:UsernameToken>
            <wsu:Timestamp xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
                <wsu:Created>2024-01-21T21:18:49Z</wsu:Created>
                <wsu:Expires>2024-01-21T21:23:49Z</wsu:Expires>
            </wsu:Timestamp>
        </wsse:Security>
    </SOAP-ENV:Header>
    <SOAP-ENV:Body>
        <Ws-info xmlns="http://www.gordic.cz/svc/xrg-ssl/v_1.0.0.0">
            <requestXml>
                <Xrg xmlns="http://www.gordic.cz/xrg/ws-info/request/v_1.0.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                    <Ws-info>
                        <Typ-pozadavku>zkraceny</Typ-pozadavku>
                    </Ws-info>
                </Xrg>
            </requestXml>
        </Ws-info>
    </SOAP-ENV:Body>
</SOAP-ENV:Envelope>`

Example of answer:
`<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" xmlns:u="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
    <s:Header>
        <o:Security s:mustUnderstand="1" xmlns:o="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
            <u:Timestamp u:Id="_0">
                <u:Created>2024-01-21T21:21:42.726Z</u:Created>
                <u:Expires>2024-01-21T21:26:42.726Z</u:Expires>
            </u:Timestamp>
        </o:Security>
    </s:Header>
    <s:Body>
        <Ws-infoResponse xmlns="http://www.gordic.cz/svc/xrg-ssl/v_1.0.0.0">
            <Ws-infoResult>
                <Xrg xmlns="http://www.gordic.cz/xrg/ws-info/response/v_1.0.0.0">
                    <Ws-info>
                        <Revize>42SSL0149018X01</Revize>
                    </Ws-info>
                </Xrg>
            </Ws-infoResult>
        </Ws-infoResponse>
    </s:Body>
</s:Envelope>`

**Comment by veewee** [comment, marked as answer]
Hallo,




> how can I solve duplicate type names in one request? Example of WSDL is https://ws.gordic.cz/390/Ssl1/Ssl.svc?WSDL and i try to implement Ws-info method, which has multiple duplicate typenames.

`Ws-info` has 3 declarations. By default, they will be combined in 1 generated class.
This is an ext-soap limitation. [More info about it and how to solve it.](https://github.com/phpro/soap-client/blob/v3.x/docs/known-issues/ext-soap.md#duplicate-typenames)
Overcoming it on a generic level requires us to write our own SOAP encoder / decoder. It's an ongoing project that probably won't be finished anytime soon. More info and how you can help out : https://github.com/phpro/soap-client/issues/485


> Second problem is, that method name is Ws-info and i got warning from line vendor/php-soap/ext-soap-engine/src/Metadata/MethodsParser.php:73, Is possible to extend regex to parsing name and return types?

I'm wondering : are you already using v3 of this package?
This should use the [wsdl-reader](https://github.com/php-soap/wsdl-reader)'s metadata parser instead of the one in the [ext-soap-engine](https://github.com/php-soap/ext-soap-engine) package. 
I did notice it takes a while to flatten this specific XML, so it might be an operation you want to perform once and use it's output as the WSDL instead:

```sh
./vendor/bin/wsdl flatten https://ws.gordic.cz/390/Ssl1/wsdl.ashx gordic.cz.wsdl
```

But to answer your question: Yes, those type-checks regexes could be extended. Care to give it a try?
It's a matter of changing these 2 methods and adding some tests for them:
https://github.com/php-soap/ext-soap-engine/blob/c1a195ff4213c3922c486078357cb3f8d5188ad5/src/Metadata/MethodsParser.php#L69-L81

**Reply by pavelwitassek** [reply]
One combined class is over, because the types are so different. I have configured remove duplicates and create class types manually, but I need help how to write TypeConverters.

I upgraded from version 2 to 3 before i try to implement this WS, I will check current version and client config again.

**Reply by veewee** [reply]
You need to alter a line in code generation in order for version 3.0 to use the new `wsdl-reader`.
You might need to check the upgrading docs:
https://github.com/phpro/soap-client/releases/tag/3.0.0


About the type-converter, It is documented here:
https://github.com/php-soap/ext-soap-engine#typeconverter

Here are some examples:
https://github.com/php-soap/ext-soap-engine/tree/main/src/Configuration/TypeConverter

You basically specify an XML namespace and a name and how you can transform from and to XML for that type.

**Reply by pavelwitassek** [reply]
I have checked the code and I am using version 3.1.1. ClientFactory use \Phpro\SoapClient\Soap\DefaultEngineFactory and code generation use \Phpro\SoapClient\Soap\CodeGeneratorEngineFactory.

**Reply by veewee** [reply]
Ah ok. So the error is during request and not during code generation.  That could indeed be the case. It will require a fix in the `ext-soap-engine` then.

**Reply by pavelwitassek** [reply]
Yes, it is during request and I try to change regex to `/^[\w-]+ (?P<name>[\w-]+)/` and `/^(?P<returnType>[\w-]+)/`. It properly parse name now, may I write tests for that and prepare PR or you do that?

**Reply by veewee** [reply]
Sure. Go ahead, all help is highly appreciated :)

**Reply by veewee** [reply]
I am working om some small improvements in the metadata generation today and decided to pick this one along with it.
It's handled in https://github.com/php-soap/ext-soap-engine/pull/21

Feel free to test it out. (tagged as 1.6)
